### PR TITLE
Overlap moe comms with collective matmul

### PR DESCRIPTION
**Pull request by Shuwen-Fang**
### Description
Implements **Overlapped All-Gather and GMM Computation (Collective Matmul)** for Ring of Experts (RoE) in MoE layers to hide communication overhead and token sorting latency. The changes include:
- Chunks input activations along the contracting embedding dimension (`embed_dim // num_chunks`). While chunk `i` computes GMM up-projections, chunk `i+1` concurrently executes Ring of Experts All-Gather and routing across expert parallel shards.
  - Update kernel to use in-place Partial Sum Accumulation to avoid unnecessary copies using `input_output_aliases`

### Testing
- **Layer Tests (`tests/unit/moe_test.py`)**: Verified chunked GMM v2 routing, partial sum accumulation, and MLP bias addition (`test_chunking_gmm_v2` with `mlp_bias=True`).
- **AOT Compilation Tests (`tests/unit/train_compile_test.py`)**: Verified ahead-of-time HLO compilation and sharding on multi-slice topologies for chunked MoE (`test_moe_chunking` and `test_moe_chunking_with_mlp_bias`).
- **Low-Level Pallas Kernel Tests (`tests/unit/pallas_mosaic_tpu_v2_kernel_test.py`)**: Verified numerical correctness and in-place buffer aliasing for partial sum accumulation across GMM (`test_gmm_partial_sum0/1`) and TGMM with zero-initialized empty expert groups (`test_tgmm_empty_group_with_partial_sum0-3`).
- **End-to-End Execution & Profiling**: Verified E2E training execution on v6e-8 TPU VM (`deepseek3-tiny` with `chunk_ag_gmm=True`, `num_chunks=4`) and generated XPlane / XProf traces confirming communication-compute overlap.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/base.yml [hunk 1]
```diff
@@ -220,6 +220,8 @@ load_balance_loss_weight: 0.0 # weight for the load balance loss
 use_random_routing: false # whether to use random routing for debug/test purpose
 use_custom_sort_vjp: true # whether to use a custom VJP sort for efficient backward pass processing in sparse matmul
 use_ring_of_experts: false # whether to use ring of experts for sparse matmul expert parallelism
+chunk_ag_gmm: false # whether to chunk All-Gather and GMM computation to overlap communication
+num_chunks: 0 # number of chunks for overlapping All-Gather and GMM computation along embedding dimension
```
Comment (NuojCheng) [inline src/maxtext/configs/base.yml]: I suggest giving them more specific names to distinguish with the other chunking feature in [cl/938756606](https://critique.corp.google.com/938756606)

File: src/maxtext/kernels/megablox/ops.py [hunk 2]
```diff
@@ -406,4 +411,4 @@ def _gmm_bwd(
   #
   # TODO(tgale, enriqueps, apaske): Fuse this transposition into the tgmm.
   drhs = drhs.swapaxes(1, 2) if transpose_rhs else drhs
-  return dlhs, drhs, None, None, grad
+  return dlhs, drhs, None, None, grad, grad if partial_sum_fwd is not None else None
```
Comment (NuojCheng) [inline src/maxtext/kernels/megablox/ops.py]: maybe move the if-else logic out of the return for better readability

File: src/maxtext/kernels/megablox/pallas_mosaic_tpu_v2_gmm_kernel.py [hunk 3]
```diff
@@ -999,7 +1001,9 @@ def validate_inputs(
   if rhs_bias is not None:
     assert rhs_bias.shape == (size_group, 1, size_n)
   if partial_sum is not None:
-    assert partial_sum.shape == (size_m, size_n)
+    assert partial_sum.shape[-1] == size_n
+    # lhs's m dimension can sometimes be padded to wi_tile_fwd_batch_seq
+    assert partial_sum.shape[0] <= size_m
```
Comment (NuojCheng) [inline src/maxtext/kernels/megablox/pallas_mosaic_tpu_v2_gmm_kernel.py:1006]: for my understanding is this change specific due to chunking feature or a general fix?

File: src/maxtext/layers/moe.py [hunk 4]
```diff
@@ -1526,59 +1556,59 @@ def route(x, logits, pre_bias_logits, rngs, input_ids=None):
...
+          x, local_sorted_indices, group_sizes, selected_experts = RoutedMoE.local_permute(
+              x,
+              global_group_sizes[None, :],
+              local_expert_size,
+              shard_index=expert_shard_id,
+              is_offset=True,
+              global_sorted_experts=selected_experts,
+              use_custom_sort_vjp=self.config.use_custom_sort_vjp,
+              use_ragged_sort=self.config.use_ragged_sort,
+          )
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py:1595]: we could move these logics to a new function named `ra2a_and_route`, mirroring your changes earlier.

File: src/maxtext/layers/moe.py [hunk 5]
```diff
@@ -1787,15 +1829,100 @@ def unsort_output_and_ra2a(intermediate_output, routing, route_metadata, output_
...
+        chunk_dim = embed_dim // self.config.num_chunks
+        chunk_rngs = (
+            jax.random.split(rngs.params(), self.config.num_chunks)
+            if self.config.use_random_routing
+            and rngs is not None
+            and hasattr(rngs, "params")
+            and callable(getattr(rngs, "params"))
+            else [rngs] * self.config.num_chunks
+        )
+        # TODO: is this valid?
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py]: add a correctness test protecting this feature for convergence if not?

File: src/maxtext/layers/moe.py [hunk 6]
```diff
@@ -1787,15 +1829,100 @@ def unsort_output_and_ra2a(intermediate_output, routing, route_metadata, output_
...
+            last_x_chunk,
+            last_w0,
+            last_w1,
+            w0_bias,
+            w1_bias,
+            gmm_fn,
+            weight_gather,
+            partial_accum0=ps0,
+            partial_accum1=ps1,
+        )
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py:1934]: I think it might be a good idea to move all the logic in if-branch to a separate function, called `moe_emb_chunking` or better names.

File: src/maxtext/layers/moe.py [hunk 7]
```diff
@@ -1787,15 +1829,100 @@ def unsort_output_and_ra2a(intermediate_output, routing, route_metadata, output_
...
+              expert_shard_id,
+              nnx.Rngs(params=chunk_rngs[chunk_idx + 1]) if chunk_rngs is not None else None,
+              input_ids=sharded_input_ids,
+          )
+          return (next_x, next_ps0, next_ps1), None
+
+        (last_x_chunk, ps0, ps1), _ = jax.lax.scan(
+            scan_fn,
+            (cur_x_chunk, partial_sum0, partial_sum1),
+            jnp.arange(self.config.num_chunks - 1),
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py]: I think you could use length=self.config.num_chunks-1 instead of arange. Also, have you tried performance when the scan is unrolled? e.g. add `unroll=self.config.num_chunks-1`, curious the performance diff.

File: tests/unit/moe_test.py [hunk 8]
```diff
@@ -537,6 +537,42 @@ def test_dense(self):
...
+        rng_hidden_states,
+        (int(cfg.per_device_batch_size) * device_count, cfg.max_target_length, cfg.base_emb_dim),
+        dtype=cfg.dtype,
+    )
+
+    devices_array = maxtext_utils.create_device_mesh(cfg)
+    mesh = Mesh(devices_array, cfg.mesh_axes)
+    variables, expected_output = self.get_expected_output(rng_model, hidden_states, cfg, mesh)
+    actual_output, _, _ = self.get_moe_output(variables, hidden_states, cfg, mesh)
+    self.assertTrue(jax.numpy.allclose(expected_output, actual_output, rtol=1e-02, atol=1e-02, equal_nan=False))
```
Comment (NuojCheng) [inline tests/unit/moe_test.py:573]: oh I see the correctness test, nice!

File: tests/unit/pallas_mosaic_tpu_v2_kernel_test.py [hunk 9]
```diff
@@ -258,6 +258,36 @@ def test_gmm_basic(self, batch_size, in_size, out_size, num_groups, has_bias, ha
 
     assert_arrays_all_close(actual, expected)
 
+  @pytest.mark.skip(reason="Test takes too long, can run locally to verify changes b/528087469")
```
Comment (NuojCheng) [inline tests/unit/pallas_mosaic_tpu_v2_kernel_test.py:261]: emmm I honestly think these tests gonna never get run....

File: tests/unit/pallas_mosaic_tpu_v2_kernel_test.py [hunk 10]
```diff
@@ -419,6 +449,51 @@ def test_tgmm_empty_group(
     self.assertEqual(actual.shape, (num_local_groups, in_size, out_size))
     assert_arrays_all_close(actual, expected)
 
+  @pytest.mark.skip(reason="Test takes too long, can run locally to verify changes b/528087469")
```
Comment (NuojCheng) [inline tests/unit/pallas_mosaic_tpu_v2_kernel_test.py:452]: emmm I honestly think these tests gonna never get run.... can we make it simple and fast?

**Review by NuojCheng — COMMENTED** [review]
Overall LGTM. Leave some comments, thank you Shuwen and great work!

File: tests/unit/pallas_mosaic_tpu_v2_kernel_test.py (hunk 10 above)
Comment (Shuwen-Fang) [reply tests/unit/pallas_mosaic_tpu_v2_kernel_test.py:452]: We are going to use tokamax kernels directly in the future instead of forking it in maxtext, so this is only going to be here temporarily

File: src/maxtext/kernels/megablox/pallas_mosaic_tpu_v2_gmm_kernel.py (hunk 3 above)
Comment (Shuwen-Fang) [reply src/maxtext/kernels/megablox/pallas_mosaic_tpu_v2_gmm_kernel.py:1006]: This occurs when embedding dim is not a multiple of wi_tile_fwd_batch_seq, so m in padded but partial sum is not. This is specifically for chunking because partial sum is only used for chunking feature

File: src/maxtext/layers/moe.py (hunk 7 above)
Comment (Shuwen-Fang) [reply src/maxtext/layers/moe.py]: That's a good idea! I'll give it a try. I'm still working on improving the performance of this feature and this PR is just to implement the feature. I can experiment & make changes if needed in a follow up PR

**Review by NuojCheng — APPROVED** [review]

File: src/maxtext/configs/types.py [hunk 11]
```diff
@@ -2591,6 +2597,26 @@ def _validate_use_te_comm_gemm_overlap(self):
           "TE Collective GEMM operations are only supported for TE quantization recipes (i.e. starting with 'te_')."
       )
 
+  def validate_use_moe_emb_chunking(self):
+    """
+    Validates that use_moe_emb_chunking is used with supported settings.
+    """
+    if self.use_moe_emb_chunking:
+      logger.warning(
```
Comment (gobbleturk) [inline src/maxtext/configs/types.py]: I would prefer to not have this warning, but up to you

File: src/maxtext/configs/base.yml [hunk 12]
```diff
@@ -220,6 +220,8 @@ load_balance_loss_weight: 0.0 # weight for the load balance loss
 use_random_routing: false # whether to use random routing for debug/test purpose
 use_custom_sort_vjp: true # whether to use a custom VJP sort for efficient backward pass processing in sparse matmul
 use_ring_of_experts: false # whether to use ring of experts for sparse matmul expert parallelism
+use_moe_emb_chunking: false # whether to chunk All-Gather and GMM computation to overlap communication
+num_moe_emb_chunks: 0 # number of chunks for overlapping All-Gather and GMM computation along embedding dimension
```
Comment (gobbleturk) [inline src/maxtext/configs/base.yml]: I would only have oneuflag - num_moe_emb_chunks - when its set to 0 its equivalent to false

File: src/maxtext/configs/types.py (hunk 11 above)
Comment (Shuwen-Fang) [reply src/maxtext/configs/types.py]: removed

File: src/maxtext/configs/base.yml (hunk 12 above)
Comment (Shuwen-Fang) [reply src/maxtext/configs/base.yml]: good idea, updated

File: src/maxtext/layers/moe.py [hunk 13]
```diff
@@ -1474,111 +1477,122 @@ def get_routed_moe_shardings(is_batch_sharded_by_expert, has_input_ids):
     ) = get_routed_moe_shardings(is_batch_sharded_by_expert, input_ids is not None)
     w0_pspec, w1_pspec, wo_pspec = maybe_aqt_partition(w0_kernel, w0_pspec, w1_kernel, w1_pspec, wo_kernel, wo_pspec)
 
-    def route(x, logits, pre_bias_logits, rngs, input_ids=None):
-      """Performs both across device and within device token routing/sorting"""
-      num_ep = self.get_expert_parallelism_size()
-      expert_shard_id = jax.lax.axis_index(self._expert_parallelism_name) if num_ep > 1 else 0
+    def roe_ag_and_route(x, logits, pre_bias_logits, num_ep, expert_shard_id, rngs, input_ids=None):
```
Comment (gobbleturk) [inline src/maxtext/layers/moe.py:1480]: Nice breakdown!

File: src/maxtext/layers/moe.py [hunk 14]
```diff
@@ -1599,6 +1613,32 @@ def route(x, logits, pre_bias_logits, rngs, input_ids=None):
           ),
       )
 
+    def route(x, logits, pre_bias_logits, rngs, input_ids=None):
```
Comment (gobbleturk) [inline src/maxtext/layers/moe.py:1616]: This looks so good!

File: src/maxtext/layers/moe.py [hunk 15]
```diff
@@ -1761,6 +1813,93 @@ def unsort_output_and_ra2a(intermediate_output, routing, route_metadata, output_
           axis_name=self._expert_parallelism_name,
       )
 
+    def moe_emb_chunking(
+        x, logits, pre_bias_logits, w0, w1, w0_bias, w1_bias, wo_bias, sharded_input_ids, rngs, embed_dim
+    ):
```
Comment (gobbleturk) [inline src/maxtext/layers/moe.py:1818]: can you include a comment for this method?

File: src/maxtext/configs/base.yml (hunk 12 above)
Comment (gobbleturk) [reply src/maxtext/configs/base.yml]: I would specify "overlapping all-gather" with "overlapping token all-gather" to distinguish it from FSDP all gathers

**Review by gobbleturk — APPROVED** [review]
